appcelerator/titanium_mobile
# feat(android)(9_3_X): add "tapjacking" prevention features

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28080

**Note:**
Tapjacking is an Android exploit where a malicious app can display a translucent system overlay on top of other apps used to intercept touches or to trick the end-user to tapping on the overlay.

**Summary:**
- Added boolean property "filterTouchesWhenObscured" to all `Ti.UI.View` derived types.
- Added event "touchfiltered" event to `Ti.UI.Button`.
  * Only fired "filterTouchesWhenObscured" is `true` and touch was obscured by overlay.
  * Intended to be used to display an alert explaining why button click was blocked.
- Added boolean property "obscured" to all touch related events.
  * Will only be true if "filterTouchesWhenObscured" is `false`.

---
**Test:**
1. Install the "AppcOverlay.apk" attached to [TIMOB-28080](https://jira.appcelerator.org/browse/TIMOB-28080) on Android.
`adb install -r <PathToApk>`
2. Build and run the below on Android.
3. Tap on "Button with Filter" and verify a "click" alert appears.
4. Verify that you see the below in the log.
```
[INFO]  @@@ touchstart obscured: false
[INFO]  @@@ touchmove obscured: false
[INFO]  @@@ touchmove obscured: false
[INFO]  @@@ touchend obscured: false
[INFO]  @@@ click obscured: false
[INFO]  @@@ singletap obscured: false
```
5. Launch the "AppcOverlay" app.
6. Tap on the "Show Fullscreen Overlay" button.
7. If "Display over other apps" window appears, then switch on, tap back, and then tap on "Fullscreen" button again.
8. A red translucent window should now fill the screen and on top of all other apps.
9. Go back to the below test app.
10. Tap on the "Button with Filter" button.
11. Verify a "touchfiltered" alert appears.
12. Tap on the "Button without Filter" button.
13. Verify a "click" alert appears.
14. Verify that you see the below in the log.
```
[INFO]  @@@ touchstart obscured: true
[INFO]  @@@ touchmove obscured: true
[INFO]  @@@ touchend obscured: true
[INFO]  @@@ click obscured: true
[INFO]  @@@ singletap obscured: true
```

app.js
```javascript
function logTouchEvent(event) {
	Ti.API.info("@@@ " + event.type + " obscured: " + event.obscured);
}
function addTouchHandlersTo(view) {
	view.addEventListener("click", function() {
		alert("click");
	});
	view.addEventListener("touchfiltered", function() {
		alert("touchfiltered\n(Overlay detected.)");
	});
	view.addEventListener("click", logTouchEvent);
	view.addEventListener("dblclick", logTouchEvent);
	view.addEventListener("singletap", logTouchEvent);
	view.addEventListener("doubletap", logTouchEvent);
	view.addEventListener("longclick", logTouchEvent);
	view.addEventListener("touchstart", logTouchEvent);
	view.addEventListener("touchmove", logTouchEvent);
	view.addEventListener("touchend", logTouchEvent);
	view.addEventListener("touchcancel", logTouchEvent);
}

var window = Ti.UI.createWindow({ title: "Tap Jacking Test" });
var button1 = Ti.UI.createButton({
	title: "Button with Filter",
	top: "25%",
	filterTouchesWhenObscured: true,
})
addTouchHandlersTo(button1);
window.add(button1);
var button2 = Ti.UI.createButton({
	title: "Button without Filter",
	bottom: "25%",
	filterTouchesWhenObscured: false,
})
addTouchHandlersTo(button2);
window.add(button2);
window.open();
```
